embarrassingly-parallel-with-delayed-or-futures.ipynb

```diff
@@ -0,0 +1,512 @@
+    "\n",
+    "This function does a simple operation: add all numbers of a list/array together, but it also sleeps for a random amount of time to simulate real work. In real use cases, this could call another python module, or even run an executable using subprocess module."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import time\n",
+    "import random\n",
+    "\n",
+    "def costly_simulation(list_param):\n",
+    "    time.sleep(random.random())\n",
+    "    return sum(list_param)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We try it locally below"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%time costly_simulation([1,2,3,4])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define the set of input parameters to call the function\n",
+    "\n",
+    "We will generate a set of inputs on which we want to run our simulation function. Here we use Pandas dataframe, but we could also use a simple list. Lets say that our simulation is run with four parameters called param_[a-d]."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "\n",
+    "input_params = pd.DataFrame(np.random.random(size=(1000, 4)),\n",
+    "                            columns=['param_a', 'param_b', 'param_c', 'param_d'])\n",
+    "input_params.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Without using Dask, we could call our simulation on all of these parameters using normal Python for loops.\n",
+    "\n",
+    "Let's only do this on a sample of our parameters as it would be quite long otherwise."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "results = []\n",
+    "for parameters in input_params.values[:10]:\n",
+    "    result = costly_simulation(parameters)\n",
+    "    results.append(result)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "results"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that this is not very clever as we can easily parallelize code. \n",
+    "\n",
+    "There are many ways to parallelize this function in Python with libraries like `multiprocessing`, `concurrent.futures`, `joblib` or others.  These are good first steps.  Dask is a good second step, especially when you want to scale across many machines.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Use Dask Delayed to make our function lazy\n",
```

Comment:
At some point in each section, it would be helpful to state when this approach should be preferred compared to the rest. For example,

> Using `dask.delayed` is a relatively straightforward way to parallelize an existing code base, even if the computation isn't embarrassingly parallel like this one. It's less useful if you need to be calling delayed functions from other delayed functions (the Futures API is better for that kind of workload). See [delayed best practices](http://dask.pydata.org/en/latest/delayed-best-practices.html) for more on using `dask.delayed` effectively.

Basically, we're presenting 3 ways of doing roughly the same thing, so to the extent possible we should give general guidelines on which one is best for common workloads.

Reply:
I see now that the Bag section does this, so perhaps my comment can be ignored.